design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+
+Another example is new functionality.  Rather than having to hack the toolkit itself, if we provide enough callbacks, developers can work on new functionality entirely outside MapLibre Native and then merge it in later if it seems important enough.  Ideally we could use that ourselves for features we may not need in a particular app, like heat maps.
+
+### Optimization
+
+_Addresses evaluation metrics [#2](#eval), [#4](#eval), [#5](#eval), and [#6](#eval)._
+
+A number of our goals are related to improving the performance of the toolkit or at least not making it much worse.  I'd propose that we try not to make it worse at all and even improve it in a few areas if the opportunity arises.
+
+For this PR let's restrict that to areas where we're likely to have made it worse and need to improve something to compensate.  For example, if we make the toolkit larger, we should try to make it smaller in another way.  If we make it slower, we'll try to make it faster to compensate.
+
+Before taking any action, the developers should generate a report on the three key areas:
+- Did the toolkit get larger on iOS or Android?
+- Did the frame rate go down on either platform?
+- Is the toolkit taking longer to startup on either platform?
+
+If the answer to any of these questions is 'yes', then we move to the next stage.
+
+#### Required Changes
+
+These are contingent on what we may have made worse and why.  Opportunities to improve that specific situation will be strategic and hard to pin down at this point.  A lot depends on why the toolkit is bigger, slower, etc...  So this is speculation right now.
+
+It depends on the other changes, but let's break it down by type.
+- Is the toolkit slower to render?  Look at the frame logic and do more up front or on other threads.
+- Is the toolkit slower to startup?  Figure out why and consider doing more work on separate threads or later.
+- Is the toolkit bigger?  Look at slicing unused layers out with compile flags, getting rid of unused code or dependencies.
+
+#### Benefits
+
+The main benefit is just keeping the performance the same.  When we implement Metal on iOS we can expect rendering performance to get much better, of course, but that's not the goal of this effort.  We aim to not make it any worse while reaching our other goals.
+
+While we're doing that, we may run across areas for improvement.  I'd suggest that the developers write these up as they find them.  Some obvious improvements are:
+- Slicing out unused functionality
+- Making more specific functionality optional
+- Reducing or eliminating dependencies on third party libraries.
+
+There's a lot that could be done in general and I'd encourage sponsors to pay maintainers to look into those.
+
+## API Modifications
+
+At present we do not expect to change any existing APIs at the public level.  We do expect to add some new API methods and classes.  We would detail these as we proceeded with the implementation.
```

Comment:
One thing that exists is MGLOpenGLStyleLayer which gives access to the openGL context. Is this something that will just remain, and then there will be a Metal one in part two, and depending on the compiled framework MGLOpenGLStyleLayer might not exist and you'd only have access to MGLMetalStyleLayer?

Reply:
A couple things I see changing with that layer.

First, I'd like to give it more information about the internal state of the renderer and the state of the toolkit so it can act more intelligently.  Right now it's got pretty minimal information.

Second yeah, we have to deal with the differences.  I think what we might do is have a superclass that's just a Custom Layer and then subclasses for the specific implementation.  We could probably even keep the MGLOpenGLStyleLayer class in place, but you'll just have to implement a new one for Metal.

Reply:
> I believe I tried shared kotlin code at the time

Like to live on the cutting edge, eh!

MGLOpenGLStyleLayer is important to lots of folks and it hints at a core idea, which is this should be easier to expand.

> Hooking into the rendering API is still a good feature to have, but there should probably be a new, more portable, custom-layer API.

I like it!  But outside the scope of this PR I think.